Enforce quota and hard cap for monthly execution minutes

tw4l opened the pull request:
Fixes #1261 Closes #1092

The quota for monthly execution minutes is treated as a hard cap. Once it is exceeded, an alert indicating that an org has exceeded its monthly execution minutes will display and the user will be unable to start new crawls. Any running crawls will be stopped once the quota is exceeded.

An execution minutes meter bar is also added in the Org Dashboard and displayed if a quota is set. More detail in https://github.com/webrecorder/browsertrix-cloud/pull/1305, which was merged into this branch.

## Changes

- Enable setting Max Execution Minutes Per Month in orgs list quotas:

<img width="535" alt="Screen Shot 2023-10-26 at 12 45 25 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/f29cd77e-3439-4428-91e0-2bb1fd88e2ff">

- Enforce quota by stopping crawls in operator once quota is reached
- Show alert banner once execution time quota is hit:

<img width="1182" alt="Screen Shot 2023-10-17 at 2 15 55 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/fd451bd9-7543-492e-a87b-960cab31e03b">

- Once quota is hit, disable Run Crawl buttons in frontend, return 403 message with `execution_minutes_quota_reached` detail in backend from crawl config `/run` endpoint, and don't run new workflows on creation (similar to storage quota)
- Display execution time for crawls in the crawl details overview, immediately below Duration:

<img width="1142" alt="Screen Shot 2023-10-17 at 2 56 24 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/2664ee07-be88-4bea-ae96-422f75fd0892">


## To test

### Manually

1. In org quotas, set the `Org Monthly Execution Minutes Quota` to a low number of minutes
2. Start a big crawl (e.g. `https://webrecorder.net` with extra hops enabled and no page limit)
3. Verify that the big crawl starts to stop once the quota has been passed
4. Verify that once the crawl finishes, it has Partial Complete status and the execution minutes alert appears at the top of the screen
6. Verify that you can no longer run crawls
7. Verify that the execution time displays in the crawl detail Overview page
8. Verify that execution minutes meter in Org Dashboard displays and is accurate

### Nightly tests

To run the nightly tests locally (requires pytest): `pytest backend/test_nightly/test_execution_minutes_quota.py`

These tests may take 5 or so minutes to complete.

tw4l committed:
First pass at enforcing execution minutes

Handled similarly to storage quotas in backend and frontend, except
execution minutes quotas are tracked and enforced monthly.

Almost surely is buggy at this stage, and there is no soft/hard cap
distinction yet.

tw4l committed:
Inc execution time in org db throughout crawl

tw4l committed:
Cancel crawl if execution mins quota is exceeded while running

tw4l committed:
Frontend fixes

tw4l committed:
Implement hard cap for execution minutes

tw4l committed:
Update execution minutes alert text

tw4l committed:
Check for execution_minutes_hard_cap_reached

tw4l committed:
Move hard cap additional minutes to org owner settings

tw4l committed:
Update field in OrgUpdateExecMinsOverage

tw4l committed:
Remove unused user arg

tw4l committed:
Fix setting overage in org owner settings

tw4l committed:
Display execution time in crawl detail overview

tw4l committed:
Fix setting crawlExecSeconds in crawl

tw4l committed:
Update docstring

tw4l committed:
Add backend test for setting org exec time overage

tw4l committed:
Fix and simplify checking quota and hard cap in org

tw4l committed:
Cancel crawl immediately when exec minutes hard cap passed

Track runningExecTime in operator CrawlStatus to track execution
time for running crawler pods. If the running crawl time + monthly
execution time tracked in org exceeds hard cap, cancel crawl.

This should cancel the crawl within about 5-10 seconds of hard cap
being exceeded (i.e. on next sync), rather than at the end of the
crawl or when a crawler pod restarts as before.

tw4l committed:
Stop rather than cancel crawl after exec mins hard cap

tw4l committed:
Add org settings billing page docs

tw4l committed:
Add nightly tests for quota and hard cap enforcement

tw4l commented:
@Shrinks99 Assigned you for a copy and docs check :)

Shrinks99 commented on docs/user-guide/org-settings.md:
```diff
@@ -24,3 +24,7 @@ Sent invites can be invalidated by pressing the trash button in the relevant _Pe
+
+## Billing
+
+This page lets organization admins set an additional number of allowed overage minutes over the monthly execution minutes quota for their organization. If set, this serves as a hard cap after which all running crawls will be stopped. When set at the default of 0, crawls will be stopped as soon as the quota is reached.
```
```suggestion
This page lets organization admins set an additional number of allowed overage minutes when the organization's monthly execution minutes quota has been reached. If set, this serves as a hard cap after which all running crawls will be stopped. When set at the default of 0, crawls will be stopped as soon as the monthly quota is reached.
```

Shrinks99 commented on frontend/src/pages/org/settings.ts:
```diff
@@ -301,6 +309,44 @@ export class OrgSettings extends LiteElement {
+      <form @submit=${this.onOrgBillingSubmit}>
+        <div class="grid grid-cols-5 gap-x-4 p-4">
+          <div class="col-span-5 md:col-span-3">
+            <sl-input
```
Would be good to add:

`inputmode="numeric"` for good keyboard selection on mobile / touch input devices

`<span slot="suffix">minutes</span>` for clarity / alignment with how we define values in the crawl workflows

tw4l replied:
Implemented

Shrinks99 commented on frontend/src/pages/org/settings.ts:
```diff
@@ -301,6 +309,44 @@ export class OrgSettings extends LiteElement {
+  private renderBilling() {
+    return html`<div class="rounded border">
+      <form @submit=${this.onOrgBillingSubmit}>
+        <div class="grid grid-cols-5 gap-x-4 p-4">
```
This needs an extra wrapper so the footer section works I think.  The footer has outer padding applied but shouldn't.

<img width="953" alt="Screenshot 2023-10-17 at 3 58 17 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5672810/7aa412d0-3f1a-4afe-926c-101bea3446e8">

See General tab for reference.
<img width="964" alt="Screenshot 2023-10-17 at 3 59 36 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/5672810/4008e8a7-6db3-43ac-926d-830c86a14e58">

tw4l replied:
Fixed!

<img width="1161" alt="Screen Shot 2023-10-17 at 5 05 16 PM" src="https://github.com/webrecorder/browsertrix-cloud/assets/6758804/fcf7dab3-e3b4-4011-95f2-e756adb95cd7">

Shrinks99 commented on frontend/src/pages/org/settings.ts:
```diff
@@ -301,6 +309,44 @@ export class OrgSettings extends LiteElement {
+            </div>
+            <div class="mt-0.5 text-xs text-neutral-500">
+              ${msg(
+                "Allowed overage minutes beyond monthly execution minutes quota before users are no longer allowed to create new crawls."
```
```suggestion
                "Allowed overage minutes beyond the organization's monthly quota. Once reached, crawl workflows will not run."
```

Shrinks99 reviewed (changes requested):
Have left a few suggestions for fixes, copy mostly good!

Once the above is dealt with, looks good to me! :D

Excited to flesh out the billing page in the future!

tw4l committed:
Apply copy changes from code review

Co-authored-by: Henry Wilkinson <henry@wilkinson.graphics>

tw4l committed:
Add inputmode numeric and minutes suffix to hard cap field

tw4l committed:
Fix Org Settings Billing page layout

tw4l commented:
@Shrinks99 ready for re-review :)

Shrinks99 commented:
Change tab title from "Billing" to "Org Limits" for now until we sort out / add more billing features to the app.

tw4l committed:
Change Billing to Limits

tw4l commented:
> Change tab title from "Billing" to "Org Limits" for now until we sort out / add more billing features to the app.

Done! I titled it "Limits" to match the other settings tabs, which don't include "Org". Also updated the backend API endpoint path for updating the setting from `/billing` to `/limits`

tw4l committed:
Update API endpoint to /limits in nightly test

tw4l commented:
This is now rebased on latest main and ready for re-review @ikreymer. I've done manual testing and re-run the nightly test to ensure all is good, and updated the issue description and docs for the Billing -> Limits tab change.

tw4l committed:
Change /billing -> /limits in backend tests

ikreymer committed:
Merge branch 'main' into issue-1261-execution-time-exceeded

ikreymer committed:
add null check

ikreymer commented:
Thanks for this! The keeping track of current running exec time is neat but has two potential issues:
- Multiple crawls may get different estimates based on when their execTime was last committed
- Month rollover will count towards the new month for long-running crawls (this is a tricky one).

I've attempted to refactor this a bit to instead force increments in short intervals to avoid tracking uncommitted exec time altogether.

ikreymer committed:
Exec Time Incremental Updates (#1314)

This adds an slight refactor to #1284 to ensure execution minutes are
updated incrementally while a crawl is running.
The execution minutes are updated on the org and on the crawl if:
- 60 seconds elapsed from last update
- any pod has newly exited
- crawl has been canceled (compute to current time)
- if month has changed

This should ensure that the execution time quota is accurate within 1
minute, and also that crawls running at the end of the month properly
count towards old and new month quotas (a bit of an edge case yes, but
bound to happen!)

other cleanup:
- add background tasks to a set to avoid premature garbage collection
(see: https://stackoverflow.com/a/74059981)
- use dt_now() consistently instead of datetime.utcnow() to store
second-rounded dates

tw4l commented:
With #1314 merged in, I think we might be finally go to go on this one!

ikreymer committed:
Issue 1261 execution time exceeded simplify (#1318)

- Remove user-set overage
- Just have a 'single execution minutes' quota, rename hard cap to be
the quota (for now)
- have backend compute storage and exec minutes quotas, report to
frontend (will eventually involve additional variables)

tw4l committed:
Rename maxCrawlMinutesPerMonth to maxExecMinutesPerMonth

tw4l committed:
Fix typing for add_crawl_config

tw4l committed:
Merge branch 'main' into issue-1261-execution-time-exceeded

tw4l committed:
Fix label for maxExecMinutesPerMonth

tw4l committed:
Allow run_now_job to be None

SuaYoo committed:
Show execution/running minutes meter in dashboard (#1305)

Co-authored-by: Tessa Walsh <tessa@bitarchivist.net>

tw4l commented:
@ikreymer @Shrinks99 Pulled in the org dashboard execution time meter into this PR, should all be ready for a last review together now

tw4l committed:
Implement code review changes

- Cap display of execution time usage at 100%
- Remove extraneous tooltip

tw4l committed:
Fix org settings Members tab

tw4l committed:
Catch new changed exec_minutes_quota_reached detail

tw4l committed:
Disable Run on Save in workflow editor when quotas are hit

Shrinks99 reviewed (approved):
Tested: Meter updating, crawl stopping gracefully after limit reached, frontend behaves properly after limit reached (crawl workflow switch included!)

LGTM! 🚀

tw4l committed:
Only go to watch tab if crawl actually started

tw4l committed:
Format with Black

tw4l committed:
this -> self

tw4l committed:
Remove messages made irrelevant by disabling run on save

ikreymer reviewed (approved):
At long last, I think this is good to go!
